### ISSUE TITLE
CPython 3.13 support (tracker issue)

### ISSUE THREAD

**Issue by neutrinoceros** [astropy/astropy#16595]
This issue is to plan and keep track on the effort to add support for CPython 3.13 .
See [PEP 719](https://peps.python.org/pep-0719/) for the release schedule.
At the time of opening, version `3.13.0b2` is available, and NumPy is already publishing nightly wheels for it.

Here's the chronological task list to complete before we can declare support for it.
- [x] pass all tests (#16659)
- [x] #16600 
- [x] #16596 
- [x] publish `cp313` wheels to PyPI (done with astropy 6.1.3)
  - [x] CPython 3.13.0rc1 needs to be released https://pythoninsider.blogspot.com/2024/08/python-3130-release-candidate-1-released.html?m=1
  - [x] update manylinux
    - [x] https://github.com/pypa/manylinux/pull/1652
    - [x] https://github.com/pypa/manylinux/pull/1660
    - [x] https://github.com/pypa/manylinux/actions/runs/10226520340
  - [x] update cibuildwheel
    - [x] https://github.com/pypa/cibuildwheel/pull/1949
    - [x] https://github.com/pypa/cibuildwheel/pull/1950
    - [x] https://github.com/pypa/cibuildwheel/releases/tag/v2.20.0
  - [x] https://github.com/OpenAstronomy/github-actions-workflows/pull/220
  - [x] cut a release or add cp313 wheels to the current latest release (the latter is more cumbersome, see https://github.com/pypa/cibuildwheel/discussions/888)
- [x] https://github.com/astropy/astropy/pull/16848
- [x] #16858 
- [ ] If we want Python 3.13 support in v6.1.x series, we also need to apply #16858 and relevant compatibility fixes to v6.1.x branch.

I'll also list which of our core dependencies already have compatible and stable wheels on PyPI, but note that it isn't a necessary condition to make our own.
 - [x] pyerfa (abi3)
 - [x] numpy v2.1.0 (2024-08-18)
 - [x] PyYAML https://github.com/yaml/pyyaml/issues/818

Supporting the experimental free-threading (a.k.a "nogil") build is a non-goal for this issue (although I note that it's currently blocked by https://github.com/liberfa/pyerfa/issues/152)
Relevant PRs are to be linked in this message.

**Comment by neutrinoceros** [comment]
#16596 currently fixes all first 3 points. The last one needs to wait until 2024-07-30 (about a month from now) when 3.13.0rc1 is (hopefully) available.

**Comment by pllim** [comment]
Maybe a little too early for a PR? According to https://peps.python.org/pep-0719/#schedule , stable release not till Oct 2024. I think we can wait for RC.

**Comment by neutrinoceros** [comment]
It is definitely earlier than previous years (because numpy is getting better at catching up with Python's development), but my reasoning is that, as always, the cost of fixing problems increases with time; if we wait for rc1 then we renounce reporting a whole category of avoidable bugs to CPython folks (i.e. bugs that aren't fixable without breaking ABI). The chance that we encounter such bugs is admittedly very small, but I still think it's worth testing now, especially considering how simple the patch currently needs to be.

In other words, my opinion is that we should start testing as soon as our dependencies allow us to.

**Comment by neutrinoceros** [comment]
Another thing to keep in mind is that Python 3.13 will be the first version to come in two flavors: GIL edition and free-threading edition. The latter is an entirely new kind of beast in the Python world and I have no idea how long it'll take for us to build compatibility for it, so the sooner we can establish we're at least compatible with the GIL build, the better.

**Comment by neutrinoceros** [comment]
Also, anecdotally, last year we found a serious memory leak while testing yt on Python 3.12 betas, which turned out to be a bug in Cython, but we only found out like 2 weeks before 3.12.0 final was scheduled. It was much more stressful because of how short a window we had to even look for it, so I don't look forward to getting this kind of late surprises this year.

**Comment by astrofrog** [comment]
I think testing as soon as our required dependencies support a prerelease seems reasonable.

**Comment by pllim** [comment]
> as soon as our required dependencies

Strictly speaking, that is only numpy.  So you might not want to use all the `devdeps`?

**Comment by neutrinoceros** [comment]
wait I think `devdeps` is indeed restricted to `pyerfa`, `numpy` and `scipy` (which is now also making cp313 wheels btw) ?

**Comment by neutrinoceros** [comment]
Oh I get it. `devdeps` indeed includes matplotlib (not ready) by default but it's really controlled by the `noscipy` factor 🙃

**Comment by pllim** [comment]
Yes, welcome to Dependency Hell™️ !

**Comment by pllim** [comment]
FWIW https://labs.quansight.org/blog/free-threaded-python-rollout

**Comment by neutrinoceros** [comment]
Aslo FYI; I just opened https://github.com/liberfa/pyerfa/pull/154, which will be needed if we plan to experiment/test against free-threaded Python sometimes this year.

**Comment by pllim** [comment]
For the item "publish cp313 wheels to PyPI" -- Is that for *after* Python 3.13 stable is out?

**Comment by neutrinoceros** [comment]
Not necessarily. The first release candidate is sufficient because it's guaranteed to be ABI stable. We should be able to provide cp313 wheels whenever our next release is.

**Comment by manodeep** [comment]
Y'all probably know about this but here is a bit of useful info on how to adapt + compile the CPython extensions for the free-threaded python - https://py-free-threading.github.io/porting/#__tabbed_1_1

Effectively add the function `PyUnstable_Module_SetGIL()` to enable GIL on free-threaded python (though GIL enabled seems to be set as the [default](https://docs.python.org/3.13/c-api/module.html#c.Py_mod_gil))

```c
PyMODINIT_FUNC
PyInit__module(void)
{
    PyObject *mod = PyModule_Create(&module);
    if (mod == NULL) {
        return NULL;
    }

#ifdef Py_GIL_DISABLED
    PyUnstable_Module_SetGIL(mod, Py_MOD_GIL_USED); 
#endif
}
```

**Comment by neutrinoceros** [comment]
Thanks @manodeep ! Though as I wrote in the OP, freethreading is out of scope for this issue. I mean to open a separate tracker issue for it soon, but it's a pretty wide topic and we could probably get a couple easy wins now.
Was it the initial motivation for your work on WCS' thread safety ?

**Comment by manodeep** [comment]
Yeah taking advantage of free-threading while not introducing data races will probably be tricky. Definitely needs a wider discussion (or waiting for a bit). 

No the motivation for WCS' thread-safety came from the reproject memory race error (that turned out to be wcs thread-race error). (Which reminds me - I need to review Mark's latest patch for the thread-safe wcs code)

**Comment by neutrinoceros** [comment]
Re:free-threading: the dedicated issue is #16916
